resource-detection/src/system/hostname.rs

```diff
@@ -10,7 +13,61 @@ pub fn get_hostname() -> Result<String, SystemDetectorError> {
 }
 
 #[cfg(target_family = "windows")]
-/// hostname getter
+/// Get the system hostname for Windows systems.
 pub fn get_hostname() -> Result<String, SystemDetectorError> {
-    unimplemented!("")
+    use std::ffi::OsString;
+    use std::os::windows::ffi::OsStringExt;
+    init();
+
+    windows_link::link!("ws2_32.dll" "system" fn GetHostNameW(name : *mut u16, namelen : i32) -> i32);
+    windows_link::link!("ws2_32.dll" "system" fn WSAGetLastError() -> i32);
+
+    // The documentation of GetHostNameW says that a buffer size of 256 is
+    // always enough.
+    let mut buffer = [0; 256];
+
+    // SAFETY: these parameters specify a valid, writable region of memory.
+    unsafe {
+        if GetHostNameW(buffer.as_mut_ptr(), buffer.len() as i32) != 0 {
+            return Err(SystemDetectorError::HostnameError(format!(
+                "GetHostNameW returned an error code: {}",
+                WSAGetLastError()
+            )));
+        }
+    }
+
+    let len = buffer
+        .iter()
+        .position(|&w| w == 0)
+        .ok_or(SystemDetectorError::HostnameError(
+            "GetHostNameW did not return a null-terminated name".to_string(),
+        ))?;
+
+    Ok(OsString::from_wide(&buffer[..len])
+        .to_string_lossy()
+        .to_string())
+}
+
+#[cfg(target_family = "windows")]
+/// Initialise the network stack for Windows.
+fn init() {
```

Comment:
> I used the same workaronud to ini the WSA as tokio/rs does

I couldn't find it. I'm a bit confused. Why do we need to bind to localhost?

Reply:
Without the init it fails with the following message:
```
Either the application has not called WSAStartup, or WSAStartup failed.
```